Custom RNG DispatchKey

pbelevich opened the pull request:
The purpose of this PR is to enable PyTorch dispatching on `at::Generator*` parameters and demonstrate how it can be used in cpp extensions to implement custom RNG.
1. `CustomRNGKeyId` value added to DispatchKey enum and `DispatchKeySet key_set_` added to `at::Generator`
2. The overloaded `operator()(at::Generator* gen)` added to MultiDispatchKeySet.
3. The existing CPUGenerator and CUDAGenerator class are supplied with CPUTensorId and CUDATensorId dispatch keys
4. The implementation of CPU's `cauchy_kernel`(as an example, because it's already moved to ATen) was templatized and moved to `ATen/native/cpu/DistributionTemplates.h` to make it available for cpp extensions
5. Minor CMake changes to make native/cpu tensors available for cpp extensions
6. RegisterCustomRNG test that demonstrates how CustomCPUGenerator class can be implemented and how custom_rng_cauchy_ native function can be registered to handle Tensor::cauchy_ calls.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32325 Custom RNG DispatchKey**

Differential Revision: [D19604558](https://our.internmc.facebook.com/intern/diff/D19604558)

pbelevich committed:
Custom RNG DispatchKey

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 74fc62e5:


**None of the build failures appear to be your fault**.

* **1/2** [broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=cd99b3706a7fd0bbeed6f2f5f99f511afa9dcfd9&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35) at merge base cd99b370 since Jan 16
  <details>
  <summary>
  Please rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch origin viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  
* **1/2** recognized as flaky :snowflake:
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=74fc62e5cf4eea54eccbd7bfff09fc88f0ee426b).

### :snowflake: 1 failure recognized as flaky
The following build failures have been detected as flaky and **may not be your fault**:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4286091) caffe2_onnx_py2_gcc5_ubuntu16_04_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=87470703) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=87470703)) :snowflake:
<details>
<summary>
<code>Jan 17 00:21:31 ./.jenkins/caffe2/test.sh: line 31:    63 Segmentation fault      (core dumped) LD_LIBRARY_PATH="$ld_library_path" "$test" --gtest_output=xml:"$gtest_reports_dir/$(basename $test).xml"</code>
</summary>

```
Jan 17 00:21:31 + for test in '$(find "$cpp_test_dir" -executable -type f)' 
Jan 17 00:21:31 + case "$test" in 
Jan 17 00:21:31 ++ basename /var/lib/jenkins/.local/lib/python2.7/site-packages/torch/test/wrapdim_test 
Jan 17 00:21:31 + LD_LIBRARY_PATH=/var/lib/jenkins/.local/lib/python2.7/site-packages/torch/lib 
Jan 17 00:21:31 + /var/lib/jenkins/.local/lib/python2.7/site-packages/torch/test/wrapdim_test --gtest_output=xml:/var/lib/jenkins/workspace/caffe2_tests/cpp/wrapdim_test.xml 
Jan 17 00:21:31 Running main() from /var/lib/jenkins/workspace/third_party/googletest/googletest/src/gtest_main.cc 
Jan 17 00:21:31 [==========] Running 1 test from 1 test case. 
Jan 17 00:21:31 [----------] Global test environment set-up. 
Jan 17 00:21:31 [----------] 1 test from TestWrapdim 
Jan 17 00:21:31 [ RUN      ] TestWrapdim.TestWrapdim 
Jan 17 00:21:31 ./.jenkins/caffe2/test.sh: line 31:    63 Segmentation fault      (core dumped) LD_LIBRARY_PATH="$ld_library_path" "$test" --gtest_output=xml:"$gtest_reports_dir/$(basename $test).xml" 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

pbelevich committed:
Update on "Custom RNG DispatchKey"




[ghstack-poisoned]

pbelevich committed:
Update on "Custom RNG DispatchKey"




[ghstack-poisoned]

ezyang commented:
I know what's going on in this PR, but others won't necessarily. Try to add more of a description to your PR.

ezyang commented on aten/src/ATen/CMakeLists.txt:
```diff
@@ -34,7 +34,7 @@ if(USE_ROCM)
 
 # NB: If you edit these globs, you'll have to update setup.py package_data as well
-FILE(GLOB base_h "*.h" "detail/*.h" "cpu/*.h")
+FILE(GLOB base_h "*.h" "detail/*.h" "cpu/*.h" "cpu/vec256/*.h" "quantized/*.h")
```
Interesting. Does this fix some recompilation problems?

ezyang replied:
It's better to put unrelated changes like this in a separate PR

pbelevich replied:
We need native/cpu/* and other headers to be available in cpp extensions because we are going to compile templatized distribution kernels

ezyang commented on aten/src/ATen/core/Generator.h:
```diff
@@ -74,6 +75,8 @@ struct CAFFE2_API Generator {
   // See Note [Acquire lock when using random generators]
   std::mutex mutex_;
 
+  virtual DispatchKeySet key_set() const = 0;
```
We should just store the DispatchKeySet inline in Generator and devirtualize this function. Ordinarily I wouldn't die on this hill, but I've had a bad experience with virtual methods on TensorImpl being impossible to remove post facto, so I'd prefer if we did it the less flexible way to start here.

ezyang commented on c10/core/DispatchKey.h:
```diff
@@ -47,6 +47,9 @@ enum class DispatchKey : uint8_t {
   SparseCPUTensorId, // PyTorch only
   SparseCUDATensorId, // PyTorch only
 
+  // Custom pseudorandom number generator
```
This comment could definitely be beefed up a little more. Also, is this a PrivateUse dispatch key, or are we allocating it specifically for secure PRNG? It's good to be clear about this, since you will be an unhappy camper if other people start using this key improperly.

ezyang replied:
This comment can still be beefed up...

ezyang commented:
Looks quite reasonable! Needs a test!

pbelevich committed:
Update on "Custom RNG DispatchKey"




[ghstack-poisoned]

pbelevich committed:
Update on "Custom RNG DispatchKey"




[ghstack-poisoned]

ezyang commented on aten/src/ATen/CPUGenerator.cpp:
```diff
@@ -51,7 +51,9 @@ CPUGenerator::CPUGenerator(uint64_t seed_in)
     next_float_normal_sample_{c10::optional<float>()},
-    next_double_normal_sample_{c10::optional<double>()} { }
+    next_double_normal_sample_{c10::optional<double>()} {
+  key_set_ = DispatchKeySet(c10::DispatchKey::CPUTensorId);
```
C++ nit: there is no reason to defer the initialization of `key_set_` into the body of the constructor, just directly initialize it as all the other fields are above

ezyang commented on aten/src/ATen/core/Generator.h:
```diff
@@ -74,6 +75,11 @@ struct CAFFE2_API Generator {
+  DispatchKeySet key_set() const { return key_set_; }
+
+  protected:
+    DispatchKeySet key_set_;
```
...and then you can make this private

ezyang reviewed (changes requested):
The WIP tag was removed but many of the requested changes have not been made.

pbelevich committed:
Update on "[WIP] Custom RNG DispatchKey"




[ghstack-poisoned]

pbelevich committed:
Update on "[WIP] Custom RNG DispatchKey"

The purpose of this PR is to enable PyTorch dispatching on `at::Generator*` parameters and demonstrate how it can be used in cpp extensions to implement custom RNG.
1. `CustomRNGKeyId` value added to DispatchKey enum and `DispatchKeySet key_set_` added to `at::Generator`
2. The overloaded `operator()(at::Generator* gen)` added to MultiDispatchKeySet.
3. The existing CPUGenerator and CUDAGenerator class are supplied with CPUTensorId and CUDATensorId dispatch keys
4. The implementation of CPU's `cauchy_kernel`(as an example, because it's already moved to ATen) was templatized and moved to `ATen/native/cpu/DistributionTemplates.h` to make it available for cpp extensions
5. Minor CMake changes to make native/cpu tensors available for cpp extensions
6. RegisterCustomRNG test that demonstrates how CustomCPUGenerator class can be implemented and how custom_rng_cauchy_ native function can be registered to handle Tensor::cauchy_ calls.




[ghstack-poisoned]

pbelevich committed:
Update on "[WIP] Custom RNG DispatchKey"

The purpose of this PR is to enable PyTorch dispatching on `at::Generator*` parameters and demonstrate how it can be used in cpp extensions to implement custom RNG.
1. `CustomRNGKeyId` value added to DispatchKey enum and `DispatchKeySet key_set_` added to `at::Generator`
2. The overloaded `operator()(at::Generator* gen)` added to MultiDispatchKeySet.
3. The existing CPUGenerator and CUDAGenerator class are supplied with CPUTensorId and CUDATensorId dispatch keys
4. The implementation of CPU's `cauchy_kernel`(as an example, because it's already moved to ATen) was templatized and moved to `ATen/native/cpu/DistributionTemplates.h` to make it available for cpp extensions
5. Minor CMake changes to make native/cpu tensors available for cpp extensions
6. RegisterCustomRNG test that demonstrates how CustomCPUGenerator class can be implemented and how custom_rng_cauchy_ native function can be registered to handle Tensor::cauchy_ calls.




[ghstack-poisoned]

ezyang commented:
Not WIP now I guess?

ezyang commented on aten/src/ATen/test/rng_test.cpp:
```diff
@@ -0,0 +1,67 @@
+  ~CustomCPUGenerator() = default;
+  uint32_t random() { return 42; }
+  uint64_t random64() { return 42; }
+  void set_current_seed(uint64_t seed) override { throw "not implemented"; }
```
I don't think this is actually valid C++ :>

ezyang commented on aten/src/ATen/test/rng_test.cpp:
```diff
@@ -0,0 +1,67 @@
+  return self;
+}
+
+template <typename T, typename RNG>
```
Err, are these copy pastes from the `template::`? Why not just call `native::templates::cauchy_kernel` directly?

ezyang reviewed (approved):
LGTM

pbelevich committed:
Update on "Custom RNG DispatchKey"

The purpose of this PR is to enable PyTorch dispatching on `at::Generator*` parameters and demonstrate how it can be used in cpp extensions to implement custom RNG.
1. `CustomRNGKeyId` value added to DispatchKey enum and `DispatchKeySet key_set_` added to `at::Generator`
2. The overloaded `operator()(at::Generator* gen)` added to MultiDispatchKeySet.
3. The existing CPUGenerator and CUDAGenerator class are supplied with CPUTensorId and CUDATensorId dispatch keys
4. The implementation of CPU's `cauchy_kernel`(as an example, because it's already moved to ATen) was templatized and moved to `ATen/native/cpu/DistributionTemplates.h` to make it available for cpp extensions
5. Minor CMake changes to make native/cpu tensors available for cpp extensions
6. RegisterCustomRNG test that demonstrates how CustomCPUGenerator class can be implemented and how custom_rng_cauchy_ native function can be registered to handle Tensor::cauchy_ calls.




[ghstack-poisoned]

pbelevich committed:
Update on "Custom RNG DispatchKey"

The purpose of this PR is to enable PyTorch dispatching on `at::Generator*` parameters and demonstrate how it can be used in cpp extensions to implement custom RNG.
1. `CustomRNGKeyId` value added to DispatchKey enum and `DispatchKeySet key_set_` added to `at::Generator`
2. The overloaded `operator()(at::Generator* gen)` added to MultiDispatchKeySet.
3. The existing CPUGenerator and CUDAGenerator class are supplied with CPUTensorId and CUDATensorId dispatch keys
4. The implementation of CPU's `cauchy_kernel`(as an example, because it's already moved to ATen) was templatized and moved to `ATen/native/cpu/DistributionTemplates.h` to make it available for cpp extensions
5. Minor CMake changes to make native/cpu tensors available for cpp extensions
6. RegisterCustomRNG test that demonstrates how CustomCPUGenerator class can be implemented and how custom_rng_cauchy_ native function can be registered to handle Tensor::cauchy_ calls.


Differential Revision: [D19604558](https://our.internmc.facebook.com/intern/diff/D19604558)

[ghstack-poisoned]

pbelevich committed:
Update on "Custom RNG DispatchKey"

The purpose of this PR is to enable PyTorch dispatching on `at::Generator*` parameters and demonstrate how it can be used in cpp extensions to implement custom RNG.
1. `CustomRNGKeyId` value added to DispatchKey enum and `DispatchKeySet key_set_` added to `at::Generator`
2. The overloaded `operator()(at::Generator* gen)` added to MultiDispatchKeySet.
3. The existing CPUGenerator and CUDAGenerator class are supplied with CPUTensorId and CUDATensorId dispatch keys
4. The implementation of CPU's `cauchy_kernel`(as an example, because it's already moved to ATen) was templatized and moved to `ATen/native/cpu/DistributionTemplates.h` to make it available for cpp extensions
5. Minor CMake changes to make native/cpu tensors available for cpp extensions
6. RegisterCustomRNG test that demonstrates how CustomCPUGenerator class can be implemented and how custom_rng_cauchy_ native function can be registered to handle Tensor::cauchy_ calls.


Differential Revision: [D19604558](https://our.internmc.facebook.com/intern/diff/D19604558)

[ghstack-poisoned]

pbelevich committed:
Update on "Custom RNG DispatchKey"

The purpose of this PR is to enable PyTorch dispatching on `at::Generator*` parameters and demonstrate how it can be used in cpp extensions to implement custom RNG.
1. `CustomRNGKeyId` value added to DispatchKey enum and `DispatchKeySet key_set_` added to `at::Generator`
2. The overloaded `operator()(at::Generator* gen)` added to MultiDispatchKeySet.
3. The existing CPUGenerator and CUDAGenerator class are supplied with CPUTensorId and CUDATensorId dispatch keys
4. The implementation of CPU's `cauchy_kernel`(as an example, because it's already moved to ATen) was templatized and moved to `ATen/native/cpu/DistributionTemplates.h` to make it available for cpp extensions
5. Minor CMake changes to make native/cpu tensors available for cpp extensions
6. RegisterCustomRNG test that demonstrates how CustomCPUGenerator class can be implemented and how custom_rng_cauchy_ native function can be registered to handle Tensor::cauchy_ calls.


Differential Revision: [D19604558](https://our.internmc.facebook.com/intern/diff/D19604558)

[ghstack-poisoned]

pbelevich committed:
Update on "Custom RNG DispatchKey"

The purpose of this PR is to enable PyTorch dispatching on `at::Generator*` parameters and demonstrate how it can be used in cpp extensions to implement custom RNG.
1. `CustomRNGKeyId` value added to DispatchKey enum and `DispatchKeySet key_set_` added to `at::Generator`
2. The overloaded `operator()(at::Generator* gen)` added to MultiDispatchKeySet.
3. The existing CPUGenerator and CUDAGenerator class are supplied with CPUTensorId and CUDATensorId dispatch keys
4. The implementation of CPU's `cauchy_kernel`(as an example, because it's already moved to ATen) was templatized and moved to `ATen/native/cpu/DistributionTemplates.h` to make it available for cpp extensions
5. Minor CMake changes to make native/cpu tensors available for cpp extensions
6. RegisterCustomRNG test that demonstrates how CustomCPUGenerator class can be implemented and how custom_rng_cauchy_ native function can be registered to handle Tensor::cauchy_ calls.


Differential Revision: [D19604558](https://our.internmc.facebook.com/intern/diff/D19604558)

[ghstack-poisoned]
